# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: layout/generic/nsBlockFrame.cpp
# Commit: 421d7bf24007
# Full Hash: 421d7bf240078ff30871183289ce376441fadb09
# Author: Jonathan Kew <jkew@mozilla.com>
# Date: 2024-11-14 21:16:19
# Description:
#   Bug 1928724 - Ensure we always have a float manager during block-frame reflow. r=layout-reviewers,emilio
#   
#   Normally we shouldn't be starting reflow from a block that is not a BFC,
#   and so BlockNeedsFloatManager would return true for the reflow root.
#   
# ==============================================================================

diff -r 19f297419354 -r 421d7bf24007 layout/generic/nsBlockFrame.cpp
--- a/layout/generic/nsBlockFrame.cpp	Thu Nov 14 10:42:38 2024 +0000
+++ b/layout/generic/nsBlockFrame.cpp	Thu Nov 14 10:44:48 2024 +0000
@@ -1470,7 +1470,16 @@
   // than keeping it around only during reflow then we should create it
   // only when there are actually floats to manage.  Otherwise things
   // like tables will gain significant bloat.
-  bool needFloatManager = nsBlockFrame::BlockNeedsFloatManager(this);
+  //
+  // See https://bugzilla.mozilla.org/show_bug.cgi?id=1931286:
+  // if we're a reflow root and no float manager is provided by the caller
+  // in aReflowInput, we'd normally expect the block to be a BFC and so
+  // BlockNeedsFloatManager will return true. But sometimes the block may
+  // have lost its BFC-ness since it was recorded as a dirty reflow root
+  // but before the reflow actually happens. Creating a float manager here
+  // avoids crashing, but may not be entirely correct in such a case.
+  bool needFloatManager =
+      !aReflowInput.mFloatManager || nsBlockFrame::BlockNeedsFloatManager(this);
   if (needFloatManager) {
     autoFloatManager.CreateFloatManager(aPresContext);
   }